services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java

```diff
@@ -85,20 +105,46 @@ public void getStream(CallContext callContext, Ticket ticket, ServerStreamListen
       return;
     }
 
-    try {
-      final CallHeaders headers = retrieveHeadersFromCallContext(callContext);
-      final UserSession session = sessionsManager.getUserSession(callContext.peerIdentity(), headers);
-      final TicketContent.PreparedStatementTicket preparedStatementTicket =
-        TicketContent.PreparedStatementTicket.parseFrom(ticket.getBytes());
+    getStreamLegacy(callContext, ticket, serverStreamListener);
+  }
+
+  private void getStreamLegacy(CallContext callContext, Ticket ticket, ServerStreamListener serverStreamListener) {
 
-      flightWorkManager.runPreparedStatement(preparedStatementTicket, serverStreamListener, allocator, session);
+    final TicketContent.PreparedStatementTicket preparedStatementTicket;
+    try {
+      preparedStatementTicket = TicketContent.PreparedStatementTicket.parseFrom(ticket.getBytes());
     } catch (InvalidProtocolBufferException ex) {
       final RuntimeException error =
         CallStatus.INVALID_ARGUMENT.withCause(ex).withDescription("Invalid ticket used in getStream")
           .toRuntimeException();
       serverStreamListener.error(error);
       throw error;
     }
+
+    UserProtos.PreparedStatementHandle preparedStatementHandle = preparedStatementTicket.getHandle();
+
+    runPreparedStatement(callContext, serverStreamListener, preparedStatementHandle);
+  }
+
+  @Override
+  public void getStreamPreparedStatement(CommandPreparedStatementQuery commandPreparedStatementQuery,
+                                         CallContext callContext, Ticket ticket,
+                                         ServerStreamListener serverStreamListener) {
+    UserProtos.PreparedStatementHandle preparedStatementHandle;
+    try {
+      preparedStatementHandle =
+        UserProtos.PreparedStatementHandle.parseFrom(commandPreparedStatementQuery.getPreparedStatementHandle());
+    } catch (InvalidProtocolBufferException e) {
+      throw CallStatus.INVALID_ARGUMENT.withDescription("Invalid PreparedStatementHandle").toRuntimeException();
+    }
+
+    // Check if given PreparedStatement is cached
+    FlightPreparedStatement preparedStatement = flightPreparedStatementCache.getIfPresent(preparedStatementHandle);
+    if (preparedStatement == null) {
+      throw CallStatus.NOT_FOUND.withDescription("PreparedStatement not found.").toRuntimeException();
```

Comment:
Should this get written to serverStreamListener, or does the caller catch this and do it?

Reply:
The caller catches it and call `listener.error` with the thrown exception, this is done on `org.apache.arrow.flight.FlightService`